File: idcloud-idv/IdCloud KYC/Helpers/Communication/KYCResponse.m
/*
MIT License

Copyright (c) 2020 Thales DIS

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

IMPORTANT: This source code is intended to serve training information purposes only.
Please make sure to review our IdCloud documentation, including security guidelines.
*/

#import "KYCResponse.h"

@implementation KYCResponse

+ (instancetype)responseWithJSON:(NSDictionary *)response {
    return [[KYCResponse alloc] initWithJSON:response];
}

- (instancetype)initWithJSON:(NSDictionary *)response {
    if (response && (self = [super init])) {
        self.code       = [response[@"code"] integerValue];
        self.message    = response[@"message"];
        self.type       = response[@"type"];
        self.document   = [KYCDocument createWithJSON:response[@"object"][@"document"]];
        self.face       = [KYCFace createWithJSON:response[@"object"][@"face"]];
    }
    
    return self;
}

- (NSString *)getMessageReadable {
    // Messages looks like "[5eb47d75-71f7-4b36-a273-8ddfe7e985bc] Internal service error". Strip down first ID part.
    NSRange loc = [_message rangeOfString:@"] "];
    if (loc.length) {
        return [_message substringFromIndex:loc.location + loc.length];
    } else {
        return _message;
    }
}

@end

File: idcloud-idv-connect/Swift/IdCloud KYC Connect/Helpers/Communication/KYCResponse.swift
/*
MIT License

Copyright (c) 2020 Thales DIS

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

IMPORTANT: This source code is intended to serve training information purposes only.
Please make sure to review our IdCloud documentation, including security guidelines.
*/

class KYCResponse: NSObject {
    private (set) var message: String?
    private (set) var type: String?
    private (set) var code: Int?
    private (set) var document: KYCDocument?
    private (set) var face: KYCFace?
    
    init?(_ json: [String: Any]!) {
        guard json != nil else {
            return nil
        }
        
        code = json["code"] as? Int ?? 0
        message = json["message"] as? String
        type = json["type"] as? String
        document = KYCDocument((json["object"] as? [String: Any])?["document"] as? [String: Any])
        
        super.init()
    }
    
    func updateWithSelfieJSON(response:[String: Any]?) {
        face = KYCFace((response?["object"] as? [String: Any])?["face"] as? [String: Any])
    }
    
    override var description : String {
        var retValue = "\(super.description)\n"
        retValue += "code: \(String(describing: code))\n"
        retValue += "message: \(String(describing: message))\n"
        retValue += "type: \(String(describing: type))\n"
        retValue += "document: \(String(describing: document))\n"
        retValue += "face: \(String(describing: face))\n"

        return retValue
    }
}
